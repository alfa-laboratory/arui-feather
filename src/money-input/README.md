```
initialState = {
    money: ''
};
function handleMoneyChange(money){
    setState({ money });
}
<MoneyInput value={ state.money } onChange={ handleMoneyChange } />
```