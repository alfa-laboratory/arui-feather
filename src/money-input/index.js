import '../input';

export default from './money-input';
