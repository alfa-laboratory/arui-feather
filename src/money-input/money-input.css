/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

@import "../vars.css";

.money-input {
    display: inline-block;
    position: relative;

    &__currency {
        display: none;
        position: absolute;
        left: 0;
        top: 10px;
        font-size: inherit;
        line-height: inherit;
        pointer-events: none;

        .input_size_s & {
            top: 7px;
        }

        .input_size_m & {
            top: 10px;
        }

        .input_size_l & {
            top: 14px;
        }

        .input_size_xl & {
            top: 17px;
        }

        .input_has-value &,
        .input_focused & {
            display: block;
        }

        .input_view_filled.input_size_s & {
            left: 8px;
            top: 0;
            line-height: 42px;
        }

        .input_view_filled.input_has-label.input_size_s & {
            top: 20px;
            line-height: inherit;
        }

        .input_view_filled.input_size_m & {
            left: 12px;
            top: 0;
            line-height: 56px;
        }

        .input_view_filled.input_has-label.input_size_m & {
            top: 28px;
            line-height: inherit;
        }
    }

    &__value {
        margin-right: 0.25em;
        color: transparent;
        font-weight: var(--font-weight-normal);
    }

    &_currency {
        .input {
            .input__addons_left {
                position: absolute;
                left: 0;
                top: 0;
                white-space: nowrap;
            }

            .input__input-wrapper:nth-child(2) {
                .input__control,
                .input__top {
                    padding-left: 0;
                }
            }
        }
    }

    &_bold {
        .input__control,
        .money-input__value {
            font-weight: var(--font-weight-bold);
        }
    }

    &_width_available {
        width: 100%;
    }
}
