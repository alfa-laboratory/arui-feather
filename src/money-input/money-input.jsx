/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/* eslint react/prop-types: 0 */

import { autobind } from 'core-decorators';
import React from 'react';
import Type from 'prop-types';

import Input from '../input/input';
import Mask from '../masked-input/mask';

import cn from '../cn';
import performance from '../performance';

const DEFAULT_FRACTION_SIZE = 2;
const DEFAULT_INTEGER_SIZE = 9;
const INTEGER_PART_SIZE = 3;

/**
 * Возвращает целую и дробную часть значения в виде массива.
 * Если дробная часть не равна `undefined`, значит введена дробная часть
 * или хотя бы запятая.
 *
 * @param {String} value Значение
 * @returns {Array.<String>}
 */
function getValueParts(value) {
    return value
        .replace(/[.бю]/g, ',') // Заменяем точки, `б` и `ю` на запятые.
        .replace(/[^\d,]/g, '') // Удаляем все, что не является цифрой или запятой.
        .split(',') // Разделяем по запятой.
        .slice(0, 2); // Отрезаем, если больше, чем один фрагмент после запятой.
}


/**
 * Сплитит интегер в группы по 3.
 *
 * @param {String} str Строка интегера
 * @returns {String}
 */
function splitInteger(str) {
    if (str.length <= INTEGER_PART_SIZE) {
        return [str];
    }

    let from = str.length - INTEGER_PART_SIZE;
    let to = str.length;

    return [str.slice(from, to)].concat(splitInteger(str.slice(0, from)));
}

/**
 * Компонент поля для ввода суммы. Может принимать в качестве значения либо число, либо число с сотой долей.
 *
 * @extends Input
 */
@cn('money-input', Input)
@performance()
class MoneyInput extends React.Component {
    static propTypes = {
        ...Input.propTypes,
        /** Максимально допустимая длина значения до запятой */
        integerLength: Type.number,
        /** Максимально допустимая длина значения после запятой */
        fractionLength: Type.number
    };

    static defaultProps = {
        fractionLength: DEFAULT_FRACTION_SIZE,
        integerLength: DEFAULT_INTEGER_SIZE
    };

    state = {
        value: ''
    };

    /**
     * @type {String}
     */
    maskPattern;

    /**
     * @type {InputMask.Pattern}
     */
    mask;

    /**
     * @type {Input}
     */
    root;

    componentWillMount() {
        this.updateMaskByValue(this.getValue());
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.value !== nextProps.value) {
            this.updateMaskByValue(nextProps.value || '');
        }
    }

    render(cn, Input) {
        return (
            <Input
                { ...this.props }
                ref={ (root) => { this.root = root; } }
                className={ cn }
                formNoValidate={ true }
                mask={ this.maskPattern }
                maxLength={ this.getMaxLength() }
                value={ this.getValue() }
                onChange={ this.handleChange }
                onProcessMaskInputEvent={ this.handleProcessMaskInputEvent }
            />
        );
    }

    @autobind
    handleProcessMaskInputEvent(event) {
        let currentValue = this.mask.format(this.getValue());
        let newValue = event.target.value;

        // При удалении отрезаем запятую, если исчезла дробная часть.
        if (newValue.length < currentValue.length) {
            let [fractionPart] = getValueParts(newValue);

            // `fractionPart !== undefined` - значит запятая введена, но
            // `fractionPart.length === 0` - значит цифр после запятой нет.
            if (fractionPart !== undefined && fractionPart.length === 0) {
                newValue = newValue.substring(0, newValue.length - 1);
                event.target.value = newValue;
            }
        }

        this.updateMaskByValue(newValue);
    }

    @autobind
    handleChange(value) {
        this.setState({ value });

        if (this.props.onChange) {
            this.props.onChange(value, Number(value.replace(/[^\d,]/g, '').replace(/,/g, '.')));
        }
    }

    /**
     * Устанавливает фокус на поле ввода.
     *
     * @public
     */
    focus() {
        this.root.focus();
    }

    /**
     * Убирает фокус с поля ввода.
     *
     * @public
     */
    blur() {
        this.root.blur();
    }

    /**
     * Скроллит страницу до поля ввода.
     *
     * @public
     */
    scrollTo() {
        this.root.scrollTo();
    }

    /**
     * Обновляет маску по значению: группирует целую часть в блоки по три символа.
     *
     * @param {String} value Значение
     */
    updateMaskByValue(value) {
        let [integerPart, fractionPart] = getValueParts(value);

        let integerPartLength = Math.max(Math.min(integerPart.length || 1, this.props.integerLength));
        this.maskPattern = splitInteger((new Array(integerPartLength + 1)).join('1')).reverse().join(' ');

        if (fractionPart !== undefined && this.props.fractionLength > 0) {
            this.maskPattern += `,${(new Array(this.props.fractionLength + 1)).join('1')}`;
        }

        this.mask = new Mask(this.maskPattern);

        if (this.root) {
            this.root.getMaskedInputInstance().setMask(this.maskPattern);
        }
    }

    /**
     * Расчитывает максимально допустимую длинну поля ввода.
     *
     * @returns {Number}
     */
    getMaxLength() {
        let maxLength = Math.floor((this.props.integerLength - 1) / INTEGER_PART_SIZE) + this.props.integerLength;

        if (this.props.fractionLength) {
            maxLength += 1 + this.props.fractionLength;
        }

        return maxLength;
    }

    /**
     * Возвращает актуальное значение для рендера.
     *
     * @returns {String}
     */
    getValue() {
        return this.props.value !== undefined ? this.props.value : this.state.value;
    }
}

export default MoneyInput;
