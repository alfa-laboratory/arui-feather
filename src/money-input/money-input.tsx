/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @ts-nocheck

/* eslint react/prop-types: 0 */

import React from 'react';
import { createCn } from 'bem-react-classname';
import { withTheme } from '../cn';

import Input, { InputProps } from '../input/input';
import Mask from '../masked-input/mask';
import { getAndroidVersion } from '../masked-input';

import { getCurrencySymbol } from '../lib/currency-codes';

const DEFAULT_FRACTION_SIZE = 2;
const DEFAULT_INTEGER_SIZE = 9;
const INTEGER_PART_SIZE = 3;

const IS_ANDROID = typeof window !== 'undefined' && /(android)/i.test(window.navigator.userAgent);

/**
 * Возвращает целую и дробную часть значения в виде массива.
 * Если дробная часть не равна `undefined`, значит введена дробная часть
 * или хотя бы запятая.
 *
 * @param value Значение
 */
function getValueParts(value: string) {
    return value
        .replace(/[.бю]/g, ',') // Заменяем точки, `б` и `ю` на запятые.
        .replace(/[^\d,]/g, '') // Удаляем все, что не является цифрой или запятой.
        .split(',') // Разделяем по запятой.
        .slice(0, 2); // Отрезаем, если больше, чем один фрагмент после запятой.
}

/**
 * Сплитит интегер в группы по 3.
 *
 * @param str Строка интегера
 */
function splitInteger(str: string): string {
    if (str.length <= INTEGER_PART_SIZE) {
        return [str];
    }

    const from = str.length - INTEGER_PART_SIZE;
    const to = str.length;

    return [str.slice(from, to)].concat(splitInteger(str.slice(0, from)));
}

type MoneyInputProps = InputProps & {
    /**
     * Максимально допустимая длина значения до запятой
     */
    integerLength?: number;
    /**
     * Максимально допустимая длина значения после запятой
     */
    fractionLength?: number;
    /**
     * Толщина шрифта
     */
    bold?: boolean;
    /**
     * Отображение символа валюты
     */
    showCurrency?: boolean;
    /**
     * Международный код валюты
     */
    currencyCode?: string;
    /**
     * Идентификатор для систем автоматизированного тестирования
     * */
    'data-test-id'?: string;
    /**
     * Обработчик события, вызываемый при событии 'change'
     * */
    onChange?: (value?: string, amount?: number) => void;
};

type MoneyInputState = {
    value: string;
}

/**
 * Компонент поля для ввода суммы. Может принимать в качестве значения либо число, либо число с сотой долей.
 */
export class MoneyInput extends React.PureComponent<MoneyInputProps, MoneyInputState> {
    protected cn = createCn('money-input');

    static defaultProps: Partial<MoneyInputProps> = {
        fractionLength: DEFAULT_FRACTION_SIZE,
        integerLength: DEFAULT_INTEGER_SIZE,
        bold: false,
        showCurrency: false,
        currencyCode: 'RUR',
    };

    state = {
        value: '',
    };

    private maskPattern: string;

    private mask: Mask;

    root: HTMLInputElement;

    private caretFixTimeout: ReturnType<typeof setTimeout> = null;

    // eslint-disable-next-line camelcase
    UNSAFE_componentWillMount() {
        this.updateMaskByValue(this.getValue());
    }

    // eslint-disable-next-line camelcase
    UNSAFE_componentWillReceiveProps(nextProps) {
        if (this.props.value !== nextProps.value) {
            this.updateMaskByValue(nextProps.value || '');
        }
    }

    componentWillUnmount() {
        clearTimeout(this.caretFixTimeout);
    }

    render() {
        return (
            <div
                className={ this.cn({
                    currency: this.props.showCurrency,
                    bold: this.props.bold,
                    width: this.props.width,
                }) }
                data-test-id={ this.props['data-test-id'] }
            >
                <Input
                    { ...this.props }
                    ref={ (root) => {
                        this.root = root;
                    } }
                    formNoValidate={ true }
                    mask={ this.maskPattern }
                    maxLength={ this.getMaxLength() }
                    value={ this.getValue() }
                    leftAddons={
                        this.props.showCurrency ? (
                            <span className={ this.cn('currency') }>
                                <span className={ this.cn('value') }>{ this.getCurrencySpacer() }</span>
                                <span>{ getCurrencySymbol(this.props.currencyCode) }</span>
                            </span>
                        ) : (
                            this.props.leftAddons
                        )
                    }
                    onChange={ this.handleChange }
                    onProcessMaskInputEvent={ this.handleProcessMaskInputEvent }
                />
            </div>
        );
    }

    private handleProcessMaskInputEvent = (event) => {
        const currentValue = this.getValue();
        const currentFormattedValue = this.mask.format(this.getValue());
        let newValue = event.target.value;
        const currentSelection = this.root.control.input.selectionStart;

        // При удалении отрезаем запятую, если исчезла дробная часть.
        if (newValue.length < currentFormattedValue.length) {
            const fractionPart = getValueParts(newValue)[1]; // Берем значение после запятой

            // `fractionPart !== undefined` - значит запятая введена, но
            // `fractionPart.length === 0` - значит цифр после запятой нет.
            if (fractionPart !== undefined && fractionPart.length === 0) {
                newValue = newValue.substring(0, newValue.length - 1);
                // eslint-disable-next-line no-param-reassign
                event.target.value = newValue;
            }
        }

        this.updateMaskByValue(newValue);

        // При добавлении последней цифры целой части каретка должна
        // оставаться перед запятой
        if (newValue.length > currentValue.length && newValue[currentSelection] === ',') {
            setTimeout((() => {
                // Фикс бага смещения каретки в браузере на андроидах Jelly Bean (c 4.1 по 4.3)
                const offsetSection = IS_ANDROID && parseFloat(getAndroidVersion() as string) < 4.4 ? 1 : 0;
                const newSelectionStart = this.root.control.input.selectionStart - 1;

                this.setInputSelection(newSelectionStart + offsetSection);
            }), 0);
        }
    };

    private handleChange = (value) => {
        this.setState({ value });

        if (this.props.onChange) {
            this.props.onChange(value, Number(value.replace(/[^\d,]/g, '').replace(/,/g, '.')));
        }
    };

    /**
     * Устанавливает фокус на поле ввода.
     */
    public focus() {
        this.root.focus();
    }

    /**
     * Убирает фокус с поля ввода.
     */
    public blur() {
        this.root.blur();
    }

    /**
     * Скроллит страницу до поля ввода.
     */
    public scrollTo() {
        this.root.scrollTo();
    }

    /**
     * Обновляет маску по значению: группирует целую часть в блоки по три символа.
     *
     * @param value Значение
     */
    private updateMaskByValue(value: string) {
        const [integerPart, fractionPart] = getValueParts(value);

        const integerPartLength = Math.max(Math.min(integerPart.length || 1, this.props.integerLength));

        this.maskPattern = splitInteger(new Array(integerPartLength + 1).join('1'))
            .reverse()
            .join(' ');

        if (fractionPart !== undefined && this.props.fractionLength > 0) {
            this.maskPattern += `,${new Array(this.props.fractionLength + 1).join('1')}`;
        }

        this.mask = new Mask(this.maskPattern);

        if (this.root) {
            this.root.getMaskedInputInstance().setMask(this.maskPattern);
        }
    }

    /**
     * Расчитывает максимально допустимую длинну поля ввода.
     */
    private getMaxLength() {
        let maxLength = Math.floor((this.props.integerLength - 1) / INTEGER_PART_SIZE) + this.props.integerLength;

        if (this.props.fractionLength) {
            maxLength += 1 + this.props.fractionLength;
        }

        return maxLength;
    }

    /**
     * Возвращает актуальное значение для рендера.
     */
    private getValue() {
        return this.props.value === undefined ? this.state.value : this.props.value;
    }

    /**
     * Устанавливает каретку поля ввода в новую позицию.
     *
     * @param selection Новое положение каретки
     */
    private setInputSelection(selection: number) {
        this.root.control.input.selectionStart = selection;
        this.root.control.input.selectionEnd = selection;
        // Android chrome имеет дефект с автоматической установкой каретки
        // при использовании клавиатуры отличной от type="text".
        if (IS_ANDROID) {
            this.setInputSelectionByTimeout(selection);
        }
    }

    /**
     * Устанавливает каретку поля ввода в заданную позицию асинхронно.
     *
     * Во-избежание дефекта с установкой каретки, наблюдаемом в мобильных браузерах,
     * установка происходит асинхронно, с минимальной задержкой,
     * с помощью [setTimeout] обертки.
     *
     * @param selection Положение каретки
     */
    private setInputSelectionByTimeout(selection: number) {
        clearTimeout(this.caretFixTimeout);

        this.caretFixTimeout = setTimeout(() => {
            this.root.control.input.selectionStart = selection;
            this.root.control.input.selectionEnd = selection;
        }, 0);
    }

    /**
     * Возвращает значение (невидимый текст) для корректного отображения значка валюты.
     */
    private getCurrencySpacer() {
        return this.mask.format(this.getValue()) || this.props.placeholder;
    }
}

class ThemedMoneyInput extends MoneyInput {}
(ThemedMoneyInput as any) = withTheme(MoneyInput);
export default ThemedMoneyInput;
