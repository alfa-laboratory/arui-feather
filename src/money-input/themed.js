import { withTheme } from '../cn';
import Component from './money-input';

export default withTheme(Component);
