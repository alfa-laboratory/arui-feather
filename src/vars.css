/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

@import './vars/font.css';
@import './vars/color.css';
@import './vars/color-theme_alfa-on-color.css';
@import './vars/color-theme_alfa-on-white.css';
@import './vars/opacity.css';
@import './vars/border-radius.css';
@import './vars/shadow.css';

:root {
    --small-screen-padding: 16px;
    --small-header-height: 80px;
    --large-screen-content-width: 1000px;
    --xlarge-screen-content-width: 1100px;
}
