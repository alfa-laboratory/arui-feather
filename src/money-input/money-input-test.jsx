/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { render, cleanUp, simulate } from '../test-utils';

import { SCROLL_TO_CORRECTION } from '../vars';

import MoneyInput from './money-input';

describe('money-input', () => {
    let originalWindowScrollTo = window.scrollTo;

    beforeEach(() => {
        window.scrollTo = sinon.spy();
    });

    afterEach(() => {
        cleanUp();
        window.scrollTo = originalWindowScrollTo;
    });

    it('should render without problems', () => {
        let moneyInput = render(<MoneyInput />);

        expect(moneyInput.node).to.exist;
        expect(moneyInput.node).to.have.class('money-input');
    });

    it('should set focused class on `focus()` call', (done) => {
        let moneyInput = render(<MoneyInput />);

        moneyInput.instance.focus();

        setTimeout(() => {
            expect(moneyInput.node).to.have.class('input_focused');
            done();
        }, 0);
    });

    it('should unset focused class on `blur()` call', (done) => {
        let moneyInput = render(<MoneyInput />);

        moneyInput.instance.focus();

        setTimeout(() => {
            moneyInput.instance.blur();

            expect(moneyInput.node).to.have.not.class('input_focused');
            done();
        }, 0);
    });

    it('should format input value passed with props', () => {
        let moneyInput = render(<MoneyInput value='1234,567' />);
        let controlNode = moneyInput.node.querySelector('input');

        expect(controlNode.value).to.equal('1 234,56');
    });

    it('should format fraction part of input value when fractionLength is 4', () => {
        let moneyInput = render(<MoneyInput value='1234,56789' fractionLength={ 4 } />);
        let controlNode = moneyInput.node.querySelector('input');

        expect(controlNode.value).to.equal('1 234,5678');
    });

    it('should call `onChange` callback after input was changed', () => {
        let onChange = sinon.spy();
        let moneyInput = render(<MoneyInput onChange={ onChange } />);
        let controlNode = moneyInput.node.querySelector('input');

        simulate(controlNode, 'change');

        expect(onChange).to.have.been.calledOnce;
    });

    it('should call `onChange` with 2 params after input was changed', () => {
        let onChange = sinon.spy();
        let moneyInput = render(<MoneyInput onChange={ onChange } />);
        let controlNode = moneyInput.node.querySelector('input');

        simulate(controlNode, 'change', { target: { value: '1 234 567,89' } });

        expect(onChange).to.have.been.calledWith('1 234 567,89', 1234567.89);
    });

    it('should scroll window to element on public scrollTo method', (done) => {
        let moneyInput = render(<MoneyInput />);
        let elemTopPosition = moneyInput.node.getBoundingClientRect().top;
        let elemScrollTo = (elemTopPosition + window.pageYOffset) - SCROLL_TO_CORRECTION;

        moneyInput.instance.scrollTo();

        setTimeout(() => {
            expect(window.scrollTo).to.have.been.calledWith(0, elemScrollTo);
            done();
        }, 0);
    });

    it('should format new value received in props', () => {
        let moneyInput = render(<MoneyInput value={ '' } />);
        let controlNode = moneyInput.node.querySelector('input');

        render(<MoneyInput value={ '1234,567' } />);

        expect(controlNode.value).to.equal('1 234,56');
    });
});
