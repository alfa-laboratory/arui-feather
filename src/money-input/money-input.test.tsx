/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import React from 'react';
import { shallow, mount } from 'enzyme';

import { SCROLL_TO_CORRECTION } from '../vars';

import { MoneyInput } from './money-input';

describe('money-input', () => {
    const originalWindowScrollTo = window.scrollTo;

    beforeEach(() => {
        window.scrollTo = jest.fn();
    });

    afterEach(() => {
        window.scrollTo = originalWindowScrollTo;
    });

    it('should render without problems', () => {
        const moneyInput = shallow(<MoneyInput />);

        expect(moneyInput).toMatchSnapshot();
    });

    it('should call input focus method on `focus()` call', () => {
        const moneyInput = mount<MoneyInput>(<MoneyInput />);
        const input = moneyInput.instance().root;

        jest.spyOn(input, 'focus');

        moneyInput.instance().focus();

        expect(input.focus).toHaveBeenCalled();
    });

    it('should call input blur method on `blur()` call', () => {
        const moneyInput = mount<MoneyInput>(<MoneyInput />);
        const input = moneyInput.instance().root;

        jest.spyOn(input, 'blur');

        moneyInput.instance().blur();

        expect(input.blur).toHaveBeenCalled();
    });

    it('should format input value passed with props', () => {
        const moneyInput = mount(<MoneyInput value="1234,567" />);
        const controlNode = moneyInput.find('input');

        expect(controlNode.prop('value')).toBe('1 234,56');
    });

    it('should format fraction part of input value when fractionLength is 4', () => {
        const moneyInput = mount(<MoneyInput value="1234,56789" fractionLength={ 4 } />);
        const controlNode = moneyInput.find('input');

        expect(controlNode.prop('value')).toBe('1 234,5678');
    });

    it('should call `onChange` callback after input was changed', () => {
        const onChange = jest.fn();
        const moneyInput = mount(<MoneyInput onChange={ onChange } />);
        const controlNode = moneyInput.find('input');

        controlNode.simulate('change');

        expect(onChange).toHaveBeenCalled();
    });

    it('should call `onChange` with 2 params after input was changed', () => {
        const onChange = jest.fn();
        const moneyInput = mount(<MoneyInput onChange={ onChange } />);
        const controlNode = moneyInput.find('input');

        controlNode.simulate('change', { target: { value: '1 234 567,89' } });

        expect(onChange).toHaveBeenCalledWith('1 234 567,89', 1234567.89);
    });

    it('should scroll window to element on public scrollTo method', () => {
        const moneyInput = mount<MoneyInput>(<MoneyInput />);
        const elemTopPosition = moneyInput.getDOMNode().getBoundingClientRect().top;
        const elemScrollTo = (elemTopPosition + window.pageYOffset) - SCROLL_TO_CORRECTION;

        moneyInput.instance().scrollTo();

        expect(window.scrollTo).toHaveBeenCalledWith(0, elemScrollTo);
    });

    it('should format new value received in props', () => {
        const moneyInput = mount(<MoneyInput value="" />);

        moneyInput.setProps({ value: '1234,567' });

        expect(moneyInput.find('input').prop('value')).toBe('1 234,56');
    });

    it('should stay caret before comma', (done) => {
        const moneyInput = mount<MoneyInput>(<MoneyInput value="12,34" />);
        const inputNode = moneyInput.find('input');

        setTimeout(() => {
            jest.useFakeTimers();

            inputNode.getDOMNode<HTMLInputElement>().selectionStart = 3;
            inputNode.getDOMNode<HTMLInputElement>().selectionEnd = 3;

            inputNode.simulate('beforeInput');
            inputNode.simulate('input', { target: { value: '123,34' } });
            jest.runAllTimers();

            expect(inputNode.getDOMNode<HTMLInputElement>().selectionStart).toBe(3);
            expect(inputNode.getDOMNode<HTMLInputElement>().selectionEnd).toBe(3);
            done();
        }, 0);
    });

    it('should use placeholder as currency spacer if value is empty', () => {
        const moneyInput = mount(<MoneyInput placeholder="placeholder" showCurrency={ true } />);
        const spacerValue = moneyInput.find('.money-input__value').text();

        expect(spacerValue).toBe('placeholder');
    });

    it('should format currency spacer the same way as value', () => {
        const moneyInput = mount(<MoneyInput value="1234,567" showCurrency={ true } />);

        const inputValue = moneyInput.find('input').prop('value');
        const spacerValue = moneyInput.find('.money-input__value').text();

        expect(spacerValue).toBe(inputValue);
    });
});
