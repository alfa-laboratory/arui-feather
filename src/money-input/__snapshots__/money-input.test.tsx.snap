// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`money-input should render without problems 1`] = `
<div
  className="money-input"
>
  <ForwardRef(ThemedComponent(Input))
    bold={false}
    currencyCode="RUR"
    formNoValidate={true}
    fractionLength={2}
    integerLength={9}
    mask="1"
    maxLength={14}
    onChange={[Function]}
    onProcessMaskInputEvent={[Function]}
    showCurrency={false}
    value=""
  />
</div>
`;
